w3lib/url.py

```diff
@@ -204,8 +204,15 @@ def _add_or_replace_parameters(url, params):
     parsed = urlsplit(url)
     args = parse_qsl(parsed.query, keep_blank_values=True)
 
-    new_args = OrderedDict(args)
-    new_args.update(params)
+    new_args = []
+    changing_params = list(params.keys())
+    for name, value in args:
+        if name in params.keys():
+            new_args.append((name, params[name]))
+            del params[name]
+        elif name not in changing_params:
+            new_args.append((name, value))
+    new_args.extend([(name, value) for name, value in params.items()])
```

Comment:
No need for list comprehension here, only `new_args.extend(params.items())` is enough.
You can use `list(params.items())` to make it explicit.
Also, `new_args += params.items()` would be more _pythonic_

Reply:
As I removed the side-effect over the input parameters, I need this list comprehension to filter the values that I don't want (the modified ones). Replaced `.extend` to `+=` anyway.